### PR TITLE
Extract big docs v2

### PR DESCRIPTION
**Pull request by PopDaph**
### Change on Core

Edit the `tokenize` route on Core to add the list of strings associated to the list of tokens. 
```
// response before
{ 
    tokens: [1844, 46110, 9424, 13576, 283, 951, 66304, 4416, 42877, 25, 17, 5163],
}

// reponse after
{ 
    tokens: [
      [1844, "Un"],
      [46110, " petit"],
      [9424, " Sou"],
      [13576, "pin"],
      [283, "ou"],
      [951, " des"],
      [66304, " bois"],
      [4416, " [["],
      [42877, "idea"],
      [25, ":"],
      [17, "2"],
      [5163, "]]"],
     ];
}
```

### Change on Front

Edit the logic to fetch the max amount of tokens for Extract using this new route. 

### Notes
- Don't review per commit 😬 
- First Rust PR 😬

### CODE REVIEW

File: core/src/providers/embedder.rs [hunk 1]
```diff
@@ -26,6 +26,8 @@ pub trait Embedder {
     async fn encode(&self, text: &str) -> Result<Vec<usize>>;
     async fn decode(&self, tokens: Vec<usize>) -> Result<String>;
 
+    async fn custom_tokenize(&self, text: String) -> Result<(Vec<usize>, Vec<String>)>;
```
Comment (PopDaph) [inline core/src/providers/embedder.rs:29]: (I decided to prefix by `custom` to make it clear that's not from tiktoken but custom for us.)

Comment (spolu) [reply core/src/providers/embedder.rs:29]: Well then I would push the custom into tiktoken :)

But really I think we can call it `tokenize` everywhere and add a small comment in tiktoken around our own functions

Comment (spolu) [reply core/src/providers/embedder.rs]: Looking

File: core/src/providers/tiktoken/tiktoken.rs [hunk 2]
```diff
@@ -241,6 +249,45 @@ impl CoreBPE {
         ret
     }
 
+    fn _tokenize_with_matching_string(&self, text: &String) -> Vec<(usize, String)> {
```
Comment (spolu) [inline core/src/providers/tiktoken/tiktoken.rs:252]: Let's add a comment to say it's our implementation

File: core/src/providers/tiktoken/tiktoken.rs [hunk 3]
```diff
@@ -241,6 +249,45 @@ impl CoreBPE {
...
+                results.push((*token, string.to_string()));
+                continue;
+            }
+
+            results.extend(Self::_tokenize_byte_pair_encode(piece, &self.encoder));
+        }
+        results
+    }
+
+    pub fn _tokenize_byte_pair_encode(
```
Comment (spolu) [inline core/src/providers/tiktoken/tiktoken.rs:269]: Let's add a comment to say it's our implementation

File: core/src/providers/tiktoken/tiktoken.rs [hunk 4]
```diff
@@ -506,6 +553,10 @@ impl CoreBPE {
         self._encode_native(text, &allowed_special).0
     }
 
+    pub fn tokenize_with_matching_string(&self, text: &String) -> Vec<(usize, String)> {
```
Comment (spolu) [inline core/src/providers/tiktoken/tiktoken.rs:556]: Let's add a comment to say it's our implementation

File: core/src/providers/tiktoken/tiktoken.rs (hunk 2 above)
Comment (spolu) [reply core/src/providers/tiktoken/tiktoken.rs:252]: Let's call it `tokenize`?

File: core/src/providers/tiktoken/tiktoken.rs (hunk 4 above)
Comment (spolu) [reply core/src/providers/tiktoken/tiktoken.rs:556]: Let's call it `tokenize`?

**Review by spolu — DISMISSED** [review]
LGTM \o/

Nice Rust!

Left minor nits re naming and comments :+1:

File: core/src/providers/tiktoken/tiktoken.rs (hunk 4 above)
Comment (PopDaph) [reply core/src/providers/tiktoken/tiktoken.rs:556]: Done!

Comment (spolu) [reply core/src/providers/tiktoken/tiktoken.rs]: You forget this one?

**Pull request merged**